index.js

```
@@ -37,41 +38,163 @@ var EchoClient = function(config){
-        if(err){
+    if (useCache === false) {
+        requestOptions.headers['cache-control'] = 'none';
+    }
+
+    this.debug(JSON.stringify(requestOptions));
+
+    request.get(requestOptions, function(err, response, body) {
+        if (err) {
             callback(err);
-        } else{
+        } else {
             callback(null, body);
         }
     });
+};
 
-    this.debug(JSON.stringify(requestOptions));
+/**
+ * Build up a query string
+ * @param {object} params
+ * @returns {string}
+ * @private
+ */
+EchoClient.prototype._queryStringParams = function(params) {
+    var queryString = '';
+    var queryStringParams = [];
+    var paramErrors = null;
+
+    var isValidParameter = function (parameter) {
+        var validParamters = [
+            'class',
+            'source',
+            'property',
+            'interval',
+            'group_by',
+            'key',
+            'value',
+            'from',
+            'to',
+            'percentile',
+            'user',
+            'filter',
+            'n'
+        ];
 
+        var hasDot = parameter.indexOf('.');
+
+        if (hasDot > -1) {
+            parameter = parameter.substring(0, hasDot);
+        }
+
+        if (validParamters.indexOf(parameter) === -1) {
+            return false;
+        } else {
+            return true;
+        }
+    }
+
+    if (!_.isEmpty(params)) {
+        for (var i in params) {
+            if (params.hasOwnProperty(i)) {
+                if (isValidParameter(i)) {
+                    queryStringParams.push(encodeURIComponent(i) + '=' + encodeURIComponent(params[i]));
+                } else {
+                    if (!paramErrors) {
+                        paramErrors = [];
+                    }
+                    paramErrors.push(i);
+                }
+            }
+        }
+        queryString += queryStringParams.join('&');
+    }
+
+    return { errors: paramErrors, queryString: queryString };
```

Why not just immediately throw an error on the first encountered param error instead of returning a bunch of them?

---

as there can be a number of params passed, it seemed sensible to accumulate the defective ones and report back to the developer by way of the log